### PR TITLE
[Part 1 of n] [SRP-23] Changes to clone action for leadership migration

### PR DESCRIPTION
**Pull request by navneet-tw**
The PR includes the following

Changes in clone action. The clone action can increase replication factor (or not). The clone action changes leadership if previous leader was in the list of source_brokers.
Added the ability in clone action to migrate to multiple target brokers based on round-robin fashion
Added the ability to restrict the migration on a topic by topic basis.
Added enough sleep times to ensure there's no overlap of executions, this is conservative, can reduce/remove this in few places later
Added some comments as notes on some changes that can be done separately( perhaps in next PR )

### CODE REVIEW

File: kafka/tools/assigner/__main__.py [hunk 1]
```diff
@@ -108,6 +109,8 @@ def main():
     tools_path = get_tools_path(args.tools_path)
     check_java_home()
 
+    time.sleep(args.pause)
```
Comment (sszp) [inline kafka/tools/assigner/__main__.py:112]: why sleep is needed here?

File: kafka/tools/assigner/__main__.py [hunk 2]
```diff
@@ -120,6 +123,7 @@ def main():
     newcluster = cluster.clone()
     action_to_run = action_map[args.action](args, newcluster)
     action_to_run.process_cluster()
+    time.sleep(args.pause)
```
Comment (sszp) [inline kafka/tools/assigner/__main__.py:126]: For all the sleeps, we should know that it is really enough!
If we are not sure about that, maybe it is better to check something else.

File: kafka/tools/assigner/actions/clone.py [hunk 3]
```diff
@@ -15,43 +15,74 @@
...
+                targeted_broker = self.cluster.brokers[to_broker]
+                if targeted_broker in partition.replicas:
+                    log.warn("Targeted broker (ID {0}) is already in the replica list for {1}:{2}"
+                             .format(targeted_broker.id, partition.topic.name, partition.num))
 
                     # If the broker is already in the replica list, it ALWAYS becomes the leader
-                    if self.to_broker != partition.replicas[0]:
-                        partition.swap_replica_positions(self.to_broker, partition.replicas[0])
+                    if targeted_broker != partition.replicas[0]:
+                        partition.swap_replica_positions(targeted_broker, partition.replicas[0])
```
Comment (sszp) [inline kafka/tools/assigner/actions/clone.py:81]: Why do we need to set the leader?

File: kafka/tools/assigner/models/reassignment.py [hunk 4]
```diff
@@ -62,9 +62,10 @@ def _execute(self, num, total, zookeeper, tools_path):
                                      '--reassignment-json-file', assignfile.name],
                                     stdout=FNULL, stderr=FNULL)
             proc.wait()
-
-            # Wait until finished
-            while True:
+            retries = 0
+            max_retries = 100
+            # Wait until finished or max retries
+            while retries < max_retries:
```
Comment (sszp) [inline kafka/tools/assigner/models/reassignment.py:68]: What happens if the retries are all used? Will it maybe try the next reassignment? Will it fail?

File: kafka/tools/assigner/__main__.py (hunk 1 above)
Comment (navneet-tw) [reply kafka/tools/assigner/__main__.py:112]: I gave a general note on sleep statements in the description, while running on staging, I will have much better sense if we really need sleep in some places or not

File: kafka/tools/assigner/__main__.py (hunk 2 above)
Comment (navneet-tw) [reply kafka/tools/assigner/__main__.py:126]: addressed above

File: kafka/tools/assigner/actions/clone.py (hunk 3 above)
Comment (navneet-tw) [reply kafka/tools/assigner/actions/clone.py:81]: this is same code as kafka-tools, just that we using multiple targets so changed variable name

Comment (navneet-tw) [reply kafka/tools/assigner/actions/clone.py:81]: basically this is the nature or clone command, if source is leader, after clone action target becomes the leader

File: kafka/tools/assigner/models/reassignment.py (hunk 4 above)
Comment (navneet-tw) [reply kafka/tools/assigner/models/reassignment.py:68]: this is just a safety net against infinite loop in the original code, in few attempts locally I saw this it was stuck in this state, we can also raise exception if retries reach max_retries , basically fail-fast

**Review by navneet-tw — COMMENTED** [review]
Replied to comments

File: kafka/tools/assigner/__main__.py (hunk 2 above)
Comment (sszp) [reply kafka/tools/assigner/__main__.py:126]: ok

File: kafka/tools/assigner/__main__.py (hunk 1 above)
Comment (sszp) [reply kafka/tools/assigner/__main__.py:112]: ok

File: kafka/tools/assigner/models/reassignment.py (hunk 4 above)
Comment (sszp) [reply kafka/tools/assigner/models/reassignment.py:68]: ok, than it is probably better to at least print something that all retries were used...

**Review by sszp — APPROVED** [review]

File: kafka/tools/assigner/models/reassignment.py (hunk 4 above)
Comment (navneet-tw) [reply kafka/tools/assigner/models/reassignment.py:68]: will throw exception so we will know that retries got exhausted

**Comment by navneet-tw** [comment]
All comments resolved/addressed

**Comment by navneet-tw** [comment]
Changes to clone action for leadership migration

**Pull request merged**